pkg/threatest/parser/main.go

```diff
@@ -0,0 +1,86 @@
+package parser
+
+import (
+	"fmt"
+	"github.com/datadog/threatest/pkg/threatest"
+	"github.com/datadog/threatest/pkg/threatest/detonators"
+	"github.com/datadog/threatest/pkg/threatest/matchers/datadog"
+	"sigs.k8s.io/yaml" // we use this library as it provides a handy "YAMLToJSON" function
+	"strings"
+	"time"
+)
+
+// Parse turns a YAML input string into a list of Threatest scenarios
+// TODO: A SSH configuration shouldn't be required at this point
+func Parse(yamlInput []byte, sshHostname string, sshUsername string, sshKey string) ([]*threatest.Scenario, error) {
+	jsonInput, err := yaml.YAMLToJSON(yamlInput)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert input YAML to JSON: %v", err)
+	}
+
+	parsed := ThreatestSchemaJson{}
+	if err := parsed.UnmarshalJSON(jsonInput); err != nil {
+		return nil, fmt.Errorf("unable to parse input: %v", err)
+	}
+
+	return buildScenarios(&parsed, sshHostname, sshUsername, sshKey)
+}
+
+func buildScenarios(parsed *ThreatestSchemaJson, sshHostname string, sshUsername string, sshKey string) ([]*threatest.Scenario, error) {
+	scenarios := []*threatest.Scenario{}
+	if len(parsed.Scenarios) == 0 {
+		return nil, fmt.Errorf("input file has no scenarios defined")
+	}
+
+	for _, parsedScenario := range parsed.Scenarios {
+		scenario := threatest.Scenario{}
+		scenario.Name = parsedScenario.Description
+
+		if parsedScenario.Detonate == nil {
+			return nil, fmt.Errorf("scenario '%s' has no detonation defined", parsedScenario.Description)
+		}
+
+		// Detonation
+		if localDetonator := parsedScenario.Detonate.LocalDetonator; localDetonator != nil {
+			commandToRun := strings.Join(parsedScenario.Detonate.LocalDetonator.Commands, "; ")
+			scenario.Detonator = detonators.NewCommandDetonator(&detonators.LocalCommandExecutor{}, commandToRun)
+		} else if remoteDetonator := parsedScenario.Detonate.RemoteDetonator; remoteDetonator != nil {
+			commandToRun := strings.Join(remoteDetonator.Commands, "; ")
+			//TODO: decouple
+			//TODO: confirm 1 SSH executor per attack makes sense
```

Comment:
This does make sense to me to offer proper isolation between scenarios executions and limit side effects 👍

Reply:
thanks for the feedback! I was mostly worried on how this would scale. If you have 50 scenarios and run them all in parallel, you'd open 50 SSH connections which would likely crash in the default setup. However I think it's reasonable since you can limit the parallelism